source/slang/slang-emit-llvm.cpp

```diff
@@ -0,0 +1,3127 @@
+                bool illegal = false;
+                for (auto field : structType->getFields())
+                {
+                    auto fieldType = field->getFieldType();
+                    auto legalizedFieldType = legalizeResourceTypes(fieldType);
+                    if (legalizedFieldType != fieldType)
+                        illegal = true;
+                }
+
+                if (illegal)
+                {
+                    IRStructType* s = irBuilder.createStructType();
+                    for (auto field : structType->getFields())
+                    {
+                        auto fieldType = field->getFieldType();
+                        auto legalizedFieldType = legalizeResourceTypes(fieldType);
+                        irBuilder.createStructField(s, field->getKey(), legalizedFieldType);
+                    }
+                    legalizedType = s;
+                }
+            }
+            break;
+        case kIROp_ArrayType:
+            {
+                auto arrayType = as<IRArrayTypeBase>(type);
+                auto elemType = arrayType->getElementType();
+                auto legalizedElemType = legalizeResourceTypes(elemType);
+                if (elemType != legalizedElemType)
+                    legalizedType =
+                        irBuilder.getArrayType(legalizedElemType, arrayType->getElementCount());
+            }
+            break;
+        default:
+            break;
+        }
+
+        legalizedResourceTypeMap[type] = legalizedType;
+        return legalizedType;
+    }
+
+    // Use this instead of the regular Slang::getOffset(), it handles
+    // legalized resource types correctly.
+    IRIntegerValue getOffset(IRStructField* field, IRTypeLayoutRules* rules)
+    {
+        IRIntegerValue offset = 0;
+        auto structType = as<IRStructType>(field->getParent());
+        auto legalStructType = as<IRStructType>(legalizeResourceTypes(structType));
+        auto legalField = field;
+        if (legalStructType != structType)
+        {
+            for (auto ff : legalStructType->getFields())
+            {
+                if (ff->getKey() == field->getKey())
+                {
+                    legalField = ff;
+                    break;
+                }
+            }
+        }
+        Slang::getOffset(*compilerOptions, rules, legalField, &offset);
+        return offset;
+    }
+
+    // Use this instead of the regular Slang::getSizeAndAlignment(), it handles
+    // legalized resource types correctly.
+    IRSizeAndAlignment getSizeAndAlignment(IRType* type, IRTypeLayoutRules* rules)
+    {
+        IRSizeAndAlignment sizeAlignment;
+
+        Slang::getSizeAndAlignment(
+            *compilerOptions,
+            rules,
+            legalizeResourceTypes(type),
+            &sizeAlignment);
+
+        return sizeAlignment;
+    }
+
+    // Returns true for any type that translates into an aggregate type in LLVM.
+    bool isAggregateType(IRType* type)
```

Comment:
This is what I was calling `isIndirectType` in my earlier comment (where I was just making up names). I think `isAggregateType` is a reasonable name since its what LLVM uses for the concept... although it is frustrating that most humans would consider a vector to be an "aggregate" type, but LLVM's categorization doesn't.